File: results/dnd-rest-benchmark/lifecycle-runs/20260730T072747Z_pi_kimi-k2p7-code_ruby-rails/app/controllers/compendium_controller.rb
class CompendiumController < ApplicationController
  def create_monster
    slug = @body['slug']
    name = @body['name']
    cr = @body['cr']
    armor_class = @body['armor_class']
    hit_points = @body['hit_points']
    tags = @body['tags']

    unless valid_id?(slug)
      bad_request('invalid slug')
      return
    end

    unless valid_non_empty_string?(name)
      bad_request('invalid name')
      return
    end

    unless valid_non_empty_string?(cr)
      bad_request('invalid cr')
      return
    end

    unless armor_class.is_a?(Integer)
      bad_request('invalid armor_class')
      return
    end

    unless hit_points.is_a?(Integer)
      bad_request('invalid hit_points')
      return
    end

    unless tags.is_a?(Array) && tags.all? { |t| t.is_a?(String) }
      bad_request('invalid tags')
      return
    end

    GameStorage.with_lock do
      begin
        GameStorage.db.execute(
          'INSERT INTO monsters (slug, name, cr, armor_class, hit_points, tags_json) VALUES (?, ?, ?, ?, ?, ?)',
          [slug, name, cr, armor_class, hit_points, JSON.generate(tags)]
        )
        render json: {
          slug: slug,
          name: name,
          cr: cr,
          armor_class: armor_class,
          hit_points: hit_points
        }, status: :created
      rescue SQLite3::ConstraintException
        render json: { error: 'slug taken' }, status: :conflict
      end
    end
  end

  def read_monster
    slug = params[:slug]
    row = GameStorage.db.get_first_row(
      'SELECT slug, name, cr, armor_class, hit_points, tags_json FROM monsters WHERE slug = ?',
      slug
    )

    if row.nil?
      render json: { error: 'monster not found' }, status: :not_found
      return
    end

    render json: {
      slug: row[0],
      name: row[1],
      cr: row[2],
      armor_class: row[3],
      hit_points: row[4],
      tags: JSON.parse(row[5])
    }
  end

  def create_item
    slug = @body['slug']
    name = @body['name']
    type = @body['type']
    rarity = @body['rarity']
    cost_gp = @body['cost_gp']

    unless valid_id?(slug)
      bad_request('invalid slug')
      return
    end

    unless valid_non_empty_string?(name)
      bad_request('invalid name')
      return
    end

    unless valid_non_empty_string?(type)
      bad_request('invalid type')
      return
    end

    unless valid_non_empty_string?(rarity)
      bad_request('invalid rarity')
      return
    end

    unless cost_gp.is_a?(Integer)
      bad_request('invalid cost_gp')
      return
    end

    GameStorage.with_lock do
      begin
        GameStorage.db.execute(
          'INSERT INTO items (slug, name, type, rarity, cost_gp) VALUES (?, ?, ?, ?, ?)',
          [slug, name, type, rarity, cost_gp]
        )
        render json: {
          slug: slug,
          name: name,
          type: type,
          rarity: rarity,
          cost_gp: cost_gp
        }, status: :created
      rescue SQLite3::ConstraintException
        render json: { error: 'slug taken' }, status: :conflict
      end
    end
  end

  def read_item
    slug = params[:slug]
    row = GameStorage.db.get_first_row(
      'SELECT slug, name, type, rarity, cost_gp FROM items WHERE slug = ?',
      slug
    )

    if row.nil?
      render json: { error: 'item not found' }, status: :not_found
      return
    end

    render json: {
      slug: row[0],
      name: row[1],
      type: row[2],
      rarity: row[3],
      cost_gp: row[4]
    }
  end
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T081334Z_claude_claude-sonnet-5_php-symfony/src/Compendium/CompendiumController.php
<?php

namespace App\Compendium;

use App\Storage\Database;
use App\Support\Json;
use App\Support\Validators;
use PDO;
use Symfony\Component\HttpFoundation\JsonResponse;

/** Handlers for the monster/item reference-data endpoints (/v1/compendium/*). */
final class CompendiumController
{
    public function createMonster(array $body): JsonResponse
    {
        $slug = $body['slug'] ?? null;
        $name = $body['name'] ?? null;
        $cr = $body['cr'] ?? null;
        $armorClass = $body['armor_class'] ?? null;
        $hitPoints = $body['hit_points'] ?? null;
        $tags = $body['tags'] ?? [];

        if (!Validators::isValidSlug($slug) || !is_string($name) || $name === ''
            || !is_string($cr) || $cr === ''
            || !Validators::isValidInt($armorClass) || !Validators::isValidInt($hitPoints)
            || !is_array($tags)) {
            return Json::error('invalid request');
        }
        foreach ($tags as $tag) {
            if (!is_string($tag)) {
                return Json::error('invalid tags');
            }
        }

        $armorClass = (int) $armorClass;
        $hitPoints = (int) $hitPoints;

        $db = Database::connection();
        $stmt = $db->prepare('SELECT slug FROM monsters WHERE slug = ?');
        $stmt->execute([$slug]);
        if ($stmt->fetchColumn() !== false) {
            return Json::error('slug already exists', 409);
        }

        $insert = $db->prepare('INSERT INTO monsters (slug, name, cr, armor_class, hit_points, tags) VALUES (?, ?, ?, ?, ?, ?)');
        $insert->execute([$slug, $name, $cr, $armorClass, $hitPoints, json_encode(array_values($tags))]);

        return new JsonResponse([
            'slug' => $slug,
            'name' => $name,
            'cr' => $cr,
            'armor_class' => $armorClass,
            'hit_points' => $hitPoints,
        ], 201);
    }

    public function getMonster(string $slug): JsonResponse
    {
        $db = Database::connection();
        $stmt = $db->prepare('SELECT slug, name, cr, armor_class, hit_points, tags FROM monsters WHERE slug = ?');
        $stmt->execute([$slug]);
        $row = $stmt->fetch(PDO::FETCH_ASSOC);
        if ($row === false) {
            return Json::error('monster not found', 404);
        }

        return new JsonResponse([
            'slug' => $row['slug'],
            'name' => $row['name'],
            'cr' => $row['cr'],
            'armor_class' => (int) $row['armor_class'],
            'hit_points' => (int) $row['hit_points'],
            'tags' => json_decode($row['tags'], true),
        ]);
    }

    public function createItem(array $body): JsonResponse
    {
        $slug = $body['slug'] ?? null;
        $name = $body['name'] ?? null;
        $type = $body['type'] ?? null;
        $rarity = $body['rarity'] ?? null;
        $costGp = $body['cost_gp'] ?? null;

        if (!Validators::isValidSlug($slug) || !is_string($name) || $name === ''
            || !is_string($type) || $type === ''
            || !is_string($rarity) || $rarity === ''
            || !Validators::isValidInt($costGp)) {
            return Json::error('invalid request');
        }

        $costGp = (int) $costGp;

        $db = Database::connection();
        $stmt = $db->prepare('SELECT slug FROM items WHERE slug = ?');
        $stmt->execute([$slug]);
        if ($stmt->fetchColumn() !== false) {
            return Json::error('slug already exists', 409);
        }

        $insert = $db->prepare('INSERT INTO items (slug, name, type, rarity, cost_gp) VALUES (?, ?, ?, ?, ?)');
        $insert->execute([$slug, $name, $type, $rarity, $costGp]);

        return new JsonResponse([
            'slug' => $slug,
            'name' => $name,
            'type' => $type,
            'rarity' => $rarity,
            'cost_gp' => $costGp,
        ], 201);
    }

    public function getItem(string $slug): JsonResponse
    {
        $db = Database::connection();
        $stmt = $db->prepare('SELECT slug, name, type, rarity, cost_gp FROM items WHERE slug = ?');
        $stmt->execute([$slug]);
        $row = $stmt->fetch(PDO::FETCH_ASSOC);
        if ($row === false) {
            return Json::error('item not found', 404);
        }

        return new JsonResponse([
            'slug' => $row['slug'],
            'name' => $row['name'],
            'type' => $row['type'],
            'rarity' => $row['rarity'],
            'cost_gp' => (int) $row['cost_gp'],
        ]);
    }
}
